src/rank.c

```diff
@@ -0,0 +1,253 @@
+  case TIES_max: vec_rank_max(v_order, v_group_sizes, n_groups, v_rank); break;
+  case TIES_sequential: vec_rank_sequential(v_order, v_group_sizes, n_groups, v_rank); break;
+  case TIES_dense: vec_rank_dense(v_order, v_group_sizes, n_groups, v_rank); break;
+  }
+
+  r_obj* out = r_null;
+  r_keep_t pi_out;
+  KEEP_HERE(out, &pi_out);
+
+  if (na_propagate) {
+    out = r_alloc_integer(size);
+    KEEP_AT(out, pi_out);
+    int* v_out = r_int_begin(out);
+    r_ssize j = 0;
+
+    for (r_ssize i = 0; i < size; ++i) {
+      v_out[i] = v_not_missing[i] ? v_rank[j++] : r_globals.na_int;
+    }
+  } else {
+    out = rank;
+  }
+
+  FREE(6);
+  return out;
+}
+
+// -----------------------------------------------------------------------------
+
+static
+void vec_rank_min(const int* v_order,
+                  const int* v_group_sizes,
+                  r_ssize n_groups,
+                  int* v_rank) {
+  r_ssize k = 0;
+  r_ssize rank = 1;
+
+  for (r_ssize i = 0; i < n_groups; ++i) {
+    const r_ssize group_size = v_group_sizes[i];
+
+    for (r_ssize j = 0; j < group_size; ++j) {
+      r_ssize loc = v_order[k] - 1;
+      v_rank[loc] = rank;
+      ++k;
+    }
+
+    rank += group_size;
+  }
+}
+
+static
+void vec_rank_max(const int* v_order,
+                  const int* v_group_sizes,
+                  r_ssize n_groups,
+                  int* v_rank) {
+  r_ssize k = 0;
+  r_ssize rank = 0;
+
+  for (r_ssize i = 0; i < n_groups; ++i) {
+    const r_ssize group_size = v_group_sizes[i];
+    rank += group_size;
+
+    for (r_ssize j = 0; j < group_size; ++j) {
+      r_ssize loc = v_order[k] - 1;
+      v_rank[loc] = rank;
+      ++k;
+    }
+  }
+}
+
+static
+void vec_rank_sequential(const int* v_order,
+                         const int* v_group_sizes,
+                         r_ssize n_groups,
+                         int* v_rank) {
+  r_ssize k = 0;
+  r_ssize rank = 1;
+
+  for (r_ssize i = 0; i < n_groups; ++i) {
+    const r_ssize group_size = v_group_sizes[i];
+
+    for (r_ssize j = 0; j < group_size; ++j) {
+      r_ssize loc = v_order[k] - 1;
+      v_rank[loc] = rank;
+      ++k;
+      ++rank;
+    }
+  }
+}
+
+static
+void vec_rank_dense(const int* v_order,
+                    const int* v_group_sizes,
+                    r_ssize n_groups,
+                    int* v_rank) {
+  r_ssize k = 0;
+  r_ssize rank = 1;
+
+  for (r_ssize i = 0; i < n_groups; ++i) {
+    const r_ssize group_size = v_group_sizes[i];
+
+    for (r_ssize j = 0; j < group_size; ++j) {
+      r_ssize loc = v_order[k] - 1;
+      v_rank[loc] = rank;
+      ++k;
+    }
+
+    ++rank;
+  }
+}
+
+// -----------------------------------------------------------------------------
+
+static inline
+enum ties parse_ties(r_obj* ties) {
+  if (r_typeof(ties) != R_TYPE_character) {
+    r_abort("`ties` must be a character vector.");
+  }
+  if (r_length(ties) < 1) {
+    r_abort("`ties` must be at least length 1.");
+  }
```

Comment:
`r_is_string()`

Reply:
Can't use that here, because the default `ties` comes in as `c("min", "max", "sequential", "dense")` and we pick the first option

Reply:
Maybe I should `arg_match0()` at the R level, and use `r_is_string()` at the C level? I like that idea

Reply:
hmm yes that makes sense. We could also export a C callable for it, that seems like a common operation in our C implementations.

Reply:
i think that would be useful